src/background/zapped.js

```diff
@@ -0,0 +1,84 @@
+/**
+ * Ghostery Browser Extension
+ * https://www.ghostery.com/
+ *
+ * Copyright 2017-present Ghostery GmbH. All rights reserved.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0
+ */
+
+import { store } from 'hybrids';
+
+import Config from '/store/config.js';
+import Options, {
+  FILTERING_MODE_GHOSTERY,
+  FILTERING_MODE_ZAP,
+} from '/store/options.js';
+
+import { FLAG_FILTERING_MODE } from '/utils/config-types.js';
+import {
+  getDynamicRulesIds,
+  PAUSED_ID_RANGE,
+  PAUSED_RULE_PRIORITY,
+  ALL_RESOURCE_TYPES,
+} from '/utils/dnr.js';
+import * as OptionsObserver from '/utils/options-observer.js';
+
+// Clear filtering mode and zapped data if the flag is removed
+store.observe(Config, (_, config, lastConfig) => {
+  if (
+    lastConfig?.hasFlag(FLAG_FILTERING_MODE) &&
+    !config.hasFlag(FLAG_FILTERING_MODE)
+  ) {
+    store.set(Options, {
+      filteringMode: FILTERING_MODE_GHOSTERY,
+      zapped: null,
+    });
+  }
+});
+
+if (__PLATFORM__ !== 'firefox') {
+  OptionsObserver.addListener(async function zapped(options, lastOptions) {
+    if (
+      options.filteringMode !== FILTERING_MODE_ZAP || // Filtering mode is not zap
+      !lastOptions || // No changes in options
+      (options.filteringMode === lastOptions.filteringMode && // Filtering mode didn't change and 'zapped' option is equal
+        OptionsObserver.isOptionEqual(options.zapped, lastOptions.zapped))
+    ) {
+      return;
+    }
+
+    const removeRuleIds = await getDynamicRulesIds(PAUSED_ID_RANGE);
+    const excludedDomains = Object.keys(options.zapped);
+
+    await chrome.declarativeNetRequest.updateDynamicRules({
+      addRules: [
+        {
+          id: 1,
+          priority: PAUSED_RULE_PRIORITY,
+          action: { type: 'allow' },
+          condition: {
+            excludedInitiatorDomains: excludedDomains,
+            excludedRequestDomains: excludedDomains,
+            resourceTypes: ALL_RESOURCE_TYPES,
+          },
+        },
+        {
+          id: 2,
+          priority: PAUSED_RULE_PRIORITY,
+          action: { type: 'allowAllRequests' },
```

Comment:
does it work as expected on Safari? Shouldn't we apply same approach we do for pausing?